src/p_mobj.c

```diff
@@ -108,6 +108,12 @@ boolean P_SetMobjState(mobj_t* mobj,statenum_t state)
   if (tempstate)
     Z_Free(tempstate);
 
+  // [FG] update object's actual height
+  if (ret)
```

Comment:
How about `if mobj->sprite != mobj->prev_sprite` so `actualheight` is only updated on sprite changes, but not frame changes?

Reply:
I tried it, and it doesn't work well with the spawn (the crosshair stays low to the ground even when it moves around).

<details>
<summary>code</summary>

```diff
diff --git a/src/p_mobj.c b/src/p_mobj.c
index 7f447183..5dd7b570 100644
--- a/src/p_mobj.c
+++ b/src/p_mobj.c
@@ -109,9 +109,10 @@ boolean P_SetMobjState(mobj_t* mobj,statenum_t state)
     Z_Free(tempstate);
 
   // [FG] update object's actual height
-  if (ret)
+  if (ret && mobj->sprite != mobj->prev_sprite)
   {
     mobj->actualheight = spritetopoffset[sprites[mobj->sprite].spriteframes[mobj->frame & FF_FRAMEMASK].lump[0]];
+    mobj->prev_sprite = mobj->sprite;
   }
 
   return ret;
@@ -891,6 +892,7 @@ mobj_t *P_SpawnMobj(fixed_t x, fixed_t y, fixed_t z, mobjtype_t type)
 
   // [FG] initialize object's actual height
   mobj->actualheight = spritetopoffset[sprites[st->sprite].spriteframes[st->frame & FF_FRAMEMASK].lump[0]];
+  mobj->prev_sprite = mobj->sprite;
 
   P_AddThinker(&mobj->thinker);
```

```diff
diff --git a/src/p_mobj.h b/src/p_mobj.h
index 7dd27dcb..4b2da3bc 100644
--- a/src/p_mobj.h
+++ b/src/p_mobj.h
@@ -382,6 +382,9 @@ typedef struct mobj_s
 
     // [FG] height of the sprite in pixels
     int actualheight;
+
+    // Update actualheight only when sprite changes.
+    spritenum_t prev_sprite;
 } mobj_t;
 
 // External declarations (fomerly in p_local.h) -- killough 5/2/98
```

</summary>

Reply:
Because of this condition in `thingheight()` I assume? 

`&& thing->actualheight > thing->height)`

Reply:
No, that doesn't affect it. I'm not sure why the spawn behaves that way. Maybe its something to do with the dehacked in DOTB.

Reply:
I still don't quite see the point of that check, though.

Reply:
It still helps when there is a sprite/hitbox mismatch like the spawn.

If direct vertical aiming is disabled, it's easy to shoot the spawn because the large `height` is used. That's the outline in this picture:

![spawn1](https://github.com/fabiangreffrath/woof/assets/56656010/5ae7a4ab-26ea-4cb3-92bf-468d8018fbd9)

If direct vertical aiming is enabled, normally it would be harder to shoot the spawn unless you're looking down at it because the calculations are done with the sprite height (left side of image). With `thing->actualheight > thing->height` added, odd cases like the spawn are handled by switching to the larger `height` and now it's much more reasonable (right side of image):

![spawn2](https://github.com/fabiangreffrath/woof/assets/56656010/f3a43e64-2ce4-4e6b-9f37-0c9be60de177)